### PR TITLE
Fix public questions with required parameters

### LINKED ISSUE #7338: Public questions with filters showing error message
- Your browser and the version: 65.0.3325.181
- Your operating system: Windows 10 64 bits
- Your databases: Postgre
- Metabase version: v0.28.5
- Metabase hosting environment: Ubuntu 16.04.4 LTS
- Metabase internal database: Mysql

Steps to reproduce the error

1)Create a question with variables

2)Make the question public

If the variables aren't set as required/mandatory, the question made public will show an error message. I think the correct way would be enable the user to set the value before the question run.

image

If the variables  are not set as required/mandatory, the question made public will run, but won’t bring the first value of the variable, only the last one.
image
image

-------

### PR DESCRIPTION
**Pull request by tlrobinson**
Resolves #7338

### CODE REVIEW

**Comment by tlrobinson** [comment]
@mazameli This only addresses #1 you mentioned in #7338. Do you think any of the others are worth doing now, or should we create new issues for them?

**Comment by tlrobinson** [comment]
@camsaul it looks like the public questions query endpoint is returning a generic string like "An error occurred" rather than a more detailed message like "Foo param is required". I'm guessing we did that to avoid leaking info. Maybe we can whitelist certain errors though?

**Comment by mazameli** [comment]
@tom I think the move would be to fix the first item in that issue, but also improve the error message; I don't think we can ship this with the current message that says "An error occurred while running the query." It should be an info message rather than a red error, and it should say e.g. "Please pick a value for **{var_1_name}** and **{{var_2_name}}**." Let me look and see if I can find prior art for this.

**Comment by tlrobinson** [comment]
This is what we show in the query builder, I assume it would be the same:

Screenshot 2019-09-10 11 21 07

**Comment by mazameli** [comment]
It should be the same in both locations, but I think we can do better than that. Give me a couple minutes to mock up something nicer.

**Comment by mazameli** [comment]
So the basic idea is to treat this as more of an expected state than an error message and use it as an opportunity to have a nice little illustration. (This illustration is super loosely about the idea of picking — like picking between an apple, an orange, and a banana.) 

question_detail

public_question

**Comment by tlrobinson** [comment]
I mentioned this in Slack, but I think this is blocked on having a better way to identify specific query errors. Currently we only get the error text, which can be dynamic (e.x. includes the param name) and can be i18n translated, so we can't even really construct regexes to match it (which is gross to begin with)

We need a stable error identifier (e.x. `required-param` in this case) in order to show the correct illustration, as well as to whitelist specific error messages for public questions.

I went ahead and pushed some temporary fixes but they will break if the error message is i18n translated.

@camsaul we should figure out how to pass error identifiers in the query response. Maybe a new `error_name` or `error_type` property alongside `error`? And how would we include that when throwing an error?

**Comment by tlrobinson** [comment]
@camsaul Can you advise how we should pass an `error_type` in the query responses?

**Comment by camsaul** [comment]
I think `:error-type`/`error_type` for now makes sense. We can add it to the `ex-info` map as `:error-type` and have the QP exception handling middleware convert it to JS-friendly `error_type`. We can also add a new `metabase.query-processor.errors` namespace or the like that has all the various error types defined so we have a definitive place to enumerate the types in the backend.

We can add additional `error-info` as well later but I don't see it being a problem having `error_type` and `error_info` both being top-level keys in an error API response

File: src/metabase/api/public.clj [hunk 1]
```diff
@@ -38,6 +38,7 @@
 (def ^:private ^:const ^Integer default-embed-max-height 800)
 (def ^:private ^:const ^Integer default-embed-max-width 1024)
 
+(def ^:private ^:const error-whitelist #"You'll need to pick a value for '.*' before this query can run\.")
```
Comment (camsaul) [inline src/metabase/api/public.clj]: This isn't a whitelist it's a single regex

File: src/metabase/api/public.clj [hunk 2]
```diff
@@ -67,7 +68,7 @@
   (card-with-uuid uuid))
 
 (defn- transform-results [results]
-  (if (= (:status results) :failed)
+  (if (and (= (:status results) :failed) (not (re-matches error-whitelist (:error results))))
```
Comment (camsaul) [inline src/metabase/api/public.clj]: I don't know what the reason we don't just return the original error message in the first place is, always returning `"An error occurred while running the query."` is not helpful. We should probably just do that instead

Comment (camsaul) [reply src/metabase/api/public.clj]: Oh wait I just realized this is for public queries. I think a key in the `ex-info` data map that lets us know the message is ok to show in embedded situtations is preferrable to a whitelist that is guaranteed to get out of sync at some point. Not sure what to call it. `:show-message-in-embeds?` or something is probably workable.

Comment (camsaul) [reply src/metabase/api/public.clj]: Also the whitelist isn't really going to work unless someone is using English

**Review by camsaul — CHANGES_REQUESTED** [review]
Generally this LGTM but instead of a error message whitelist I think something like a `:show-message-in-embeds?` key that gets thrown along with the `ex-info` is harder to mess up and works with non-English languages.

Also it would be nice to add tests to make sure the new `:show-message-in-embeds?` key is respecte, a single test for `metabase.api.public/transform-results` is probably sufficient

File: src/metabase/api/public.clj (hunk 2 above)
Comment (tlrobinson) [reply src/metabase/api/public.clj]: Yes, I mentioned this was a temporary solution:

> I went ahead and pushed some temporary fixes but they will break if the error message is i18n translated.

**Comment by tlrobinson** [comment]
@camsaul My proposal was to include an `error_id` (or whatever) via `ex-info`. We would then use that in both the public questions error whitelist and on the frontend to show custom UI for specific errors.

IMHO that's better than `show-message-in-embeds?` since it centralizes the list of allowed error messages in the public code.

Can you implement that?

**Comment by camsaul** [comment]
Ok @tlrobinson I made the changes here for the backend stuff. I'll probably need to update a few tests still but I'll do that tomorrow

**Comment by tlrobinson** [comment]
@camsaul Cool thanks. Last two things:

1. the key in the response should be `error_type` instead of `error-type`
2. for public embeds we need to return the same `{ status, error, error_type }` object as normal, so the frontend can get both the `error` message and `error_type`. Currently it's just returning the string.

I tried making this change but it doesn't seem to be working quite right:

```diff
diff --git a/src/metabase/api/public.clj b/src/metabase/api/public.clj
index 9f53818b3..32f054556 100644
--- a/src/metabase/api/public.clj
+++ b/src/metabase/api/public.clj
@@ -75,16 +75,15 @@
   [results]
   (u/select-nested-keys
    results
-   [[:data :cols :rows :rows_truncated :insights] [:json_query :parameters] :error :status]))
+   [[:data :cols :rows :rows_truncated :insights] [:json_query :parameters] :error :error_type :status]))
 
 (defmethod transform-results :failed
-  [{:keys [error error-type], :as results}]
+  [{:keys [error error_type error-type]}]
   ;; if the query failed instead, unless the error type is specified and is EXPLICITLY allowed to be shown for embeds,
   ;; instead of returning anything about the query just return a generic error message
-  (let [message (if (qp.error-type/show-in-embeds? error-type)
-                  error
-                  (tru "An error occurred while running the query."))]
-    (ex-info message {:status-code 400})))
+  (if (qp.error-type/show-in-embeds? error_type)
+    {:status :failed, :error error, :error_type error_type }
+    {:status :failed, :error (tru "An error occurred while running the query.") }))
 
 (defn run-query-for-card-with-id-async
   "Run the query belonging to Card with `card-id` with `parameters` and other query options (e.g. `:constraints`).
diff --git a/src/metabase/query_processor/middleware/catch_exceptions.clj b/src/metabase/query_processor/middleware/catch_exceptions.clj
index 2c6ff4b40..35a6cd3c6 100644
--- a/src/metabase/query_processor/middleware/catch_exceptions.clj
+++ b/src/metabase/query_processor/middleware/catch_exceptions.clj
@@ -88,7 +88,7 @@
                                (explain-schema-validation-error error))]
        {:error error-msg})
      (when (qp.error-type/known-error-type? error-type)
-       {:error-type error-type})
+       {:error_type error-type})
      {:ex-data (dissoc data :schema)})))
 
 
```

**Pull request merged**